### PR TITLE
[ADD] Dotnet libraries V0

### PR DESCRIPTION
**Pull request by mathi123**
First version of OdooC# repository.
v1:
- Annotate C# classes, referring to Odoo models
- Annotate C# properties, referring to Odoo fields, One2many fields and Many2one fields.
- Read/Write/update classes -> odoo rows get updated.

Todo:
- Lazy load One2many and Many2one properties automagically

**Review Guidelines**
- This repository is a visual studio solution.
- The only external reference of this solution is NUnit testing library. You can use nuget restore, to restore in in visual studio.
- xmlrpc lib source is included, but does _not_ need review!
- OdooRpcWrapper is the core library who transforms the attribute annotations in XMLRpc structs, this library does need review. 
- OdooTypedClasses is a class library containing some example classes that you can use or inherit from, you can also configure your own classes from scratch.
- For examples, see Readme and unit tests.

### CODE REVIEW

**Comment by mathi123** [comment]
@bwrsandman can you create needs-review label for this repo?

**Comment by bwrsandman** [comment]
Added labels and enabled tests

**Comment by gdgellatly** [comment]
I've mailed the link to this to a couple of .net people.  I am not sure, but maybe from a review perspective to reduce the diff, it would be easier to have an initial base merge of stuff that doesn't need review

e.g. .git*, cook xmlrpc

which is about 70%+ of all the code.  I don't know just an idea, maybe c# people know what to ignore.

**Comment by mathi123** [comment]
@gdgellatly thanks! I have included some review guidelines above, or would you recommend merging the xmlrpc lib already?

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      - python-simplejson
+      - python-serial
+
+env:
+  matrix:
+  - VERSION="8.0"
+
+install:
+  - createdb test
+  - git clone https://github.com/mathi123/maintainer-quality-tools.git ${HOME}/maintainer-quality-tools -b unhardcode
```
Comment (bwrsandman) [inline .travis.yml]: You can use the OCA MQT now

File: OdooRpcWrapper/IOdooObjectRpc.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿using CookComputing.XmlRpc;
+using System;
+
+namespace OdooRpcWrapper
+{
+    [XmlRpcUrl("object")]
+    public interface IOdooObjectRpc : IXmlRpcProxy
+    {
+        [XmlRpcMethod("execute")]
+        int create(String dbName, int userId, string dbPwd, string model, string method, XmlRpcStruct fieldValues);
```
Comment (DamonLiu0814) [inline OdooRpcWrapper/IOdooObjectRpc.cs:10]: s/String/string/

File: OdooRpcWrapper/OdooAPI.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                loginRpc.Proxy = _networkProxy;
+            }
+
+            if (_certificateValidationFunction != null)
+            {
+                ServicePointManager.ServerCertificateValidationCallback = _certificateValidationFunction;
+            }
+            else
+            {
+                ServicePointManager.ServerCertificateValidationCallback = StandardValidationFunction;
```
Comment (DamonLiu0814) [inline OdooRpcWrapper/OdooAPI.cs:47]: StandardValidationFunction ?

File: OdooRpcWrapper/OdooAPI.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            }
+
+            // Log in and get user id
+            object result = null;
+            try
+            {
+                result = loginRpc.login(_credentials.DbName, _credentials.DbUser, _credentials.DbPassword);
+
+                if (result is bool)
```
Comment (DamonLiu0814) [inline OdooRpcWrapper/OdooAPI.cs:56]: I think use 
switch( resule.GetType() )
{
   case bool:
     .......
} is better and easy to read

File: OdooRpcWrapper/OdooAPI.cs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+                {
+                    // Login failed!
+                    _credentials.UserId = -1;
+                    _loginResult = OdooLoginResult.InvalidCredentials;
+                }
+                else if (result is int)
```
Comment (DamonLiu0814) [inline OdooRpcWrapper/OdooAPI.cs:62]: Don't use **else if** if it's possible . It's hard to read when the code is too long

File: OdooRpcWrapper/OdooAPI.cs [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+                {
+                    _credentials.UserId = (int)result;
```
Comment (DamonLiu0814) [inline OdooRpcWrapper/OdooAPI.cs:64]: when you defind result , you can use **var** , so  you don't need  **(int)result**

File: OdooRpcWrapper/OdooAPI.cs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                _loginResult = OdooLoginResult.Unknown;
+            }
+
+            return _loginResult;
+        }
+
+        private bool StandardValidationFunction(object sender, X509Certificate certificate, 
+            X509Chain chain, SslPolicyErrors sslPolicyErrors)
+        {
+            return true;
```
Comment (DamonLiu0814) [inline OdooRpcWrapper/OdooAPI.cs:99]: this function seems  not implemented?

**Review by DamonLiu0814 — COMMENTED** [review]
too many file need review , maybe you can ping other C# coder to review .

**Pull request closed without merging**